comet-java-client/src/main/java/ml/comet/experiment/impl/BaseExperiment.java

```diff
@@ -0,0 +1,899 @@
+     *                    Can be {@code null} if not interested in completion signal.
+     */
+    void logMetricAsync(@NonNull String metricName, @NonNull Object metricValue,
+                        long step, long epoch, Action onComplete) {
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug("logMetricAsync {} = {}, step: {}, epoch: {}", metricName, metricValue, step, epoch);
+        }
+
+        MetricRest metricRequest = createLogMetricRequest(metricName, metricValue, step, epoch, this.context);
+        this.sendAsynchronously(restApiClient::logMetric, metricRequest, onComplete);
+    }
+
+    /**
+     * Synchronous version that waits for result or exception. Also, it checks the response status for failure.
+     *
+     * @param parameterName The name of the param being logged
+     * @param paramValue    The value for the param being logged
+     * @param step          The current step for this metric, this will set the given step for this experiment
+     */
+    @Override
+    public void logParameter(@NonNull String parameterName, @NonNull Object paramValue, long step) {
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug("logParameter {} = {}, step: {}", parameterName, paramValue, step);
+        }
+
+        sendSynchronously(restApiClient::logParameter,
+                createLogParamRequest(parameterName, paramValue, step, this.context));
+    }
+
+    /**
+     * Asynchronous version that only logs any received exceptions or failures.
+     *
+     * @param parameterName The name of the param being logged
+     * @param paramValue    The value for the param being logged
+     * @param step          The current step for this metric, this will set the given step for this experiment
+     * @param onComplete    The optional action to be invoked when this operation asynchronously completes.
+     *                      Can be {@code null} if not interested in completion signal.
+     */
+    void logParameterAsync(@NonNull String parameterName, @NonNull Object paramValue, long step, Action onComplete) {
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug("logParameterAsync {} = {}, step: {}", parameterName, paramValue, step);
+        }
+
+        ParameterRest paramRequest = createLogParamRequest(parameterName, paramValue, step, this.context);
+        this.sendAsynchronously(restApiClient::logParameter, paramRequest, onComplete);
+    }
+
+    /**
+     * Synchronous version that waits for result or exception. Also, it checks the response status for failure.
+     *
+     * @param line   Text to be logged
+     * @param offset Offset describes the place for current text to be inserted
+     * @param stderr the flag to indicate if this is StdErr message.
+     */
+    @Override
+    public void logLine(String line, long offset, boolean stderr) {
+        validate();
```

Comment:
Yes, the `logLine` is special in terms that it should not output anything at least in our code to avoid infinite recursion.

Reply:
I'm confused why we are checking if the experiment key is empty in `logLine`, when could this happens?

As for checking for if the experiment is alive, it can happens if the experiment has been closed and we try to log a line. In that case, it seems we are showing a traceback to the user, could you confirm that? https://github.com/comet-ml/comet-java-sdk/pull/37/files#diff-118556a8d152628c27cc4bbe43395fec2d7a0fd05bd29d257c726982c2e6c74aR108. I don't think the behavior has changed, I just want to know if I need to create a cleaning ticket.

Reply:
This is a validation of the internal state of the experiment. It is here for a reason. You are right that situation when experiment key is missing at this point is not a part of normal flow. However, this can happen due some bugs, etc. and with this check we guarantee that when we go further the state of application as expected. Or it fails in predictable manner with clear message.

I believe it does not worth such a long discussion at all.